### DISCUSSION TITLE
Code using nodes instead of  graph

### DISCUSSION THREAD

**Discussion by omar92** [Comfy-Org/ComfyUI#6284] [Ideas]
the idea that i wish to discuss if there are a possible way to use comfy similar to following 
![image](https://github.com/user-attachments/assets/e129496e-4fe3-4a20-9eca-cf98740ac60f)

instead of drawing that graph we would write 

```
 LoadCheckpoint("modelaName", OnModelLoaded );

Function  OnModelLoaded (model , clip , vae){

ClipText("positive prompt" , (posCondition)=>{

  ClipText("positive prompt" , (negCondition)=>{
  
    var seed = 331231231;
     //.... etc
    Ksampler(seed, .... , (Latent)=>{

        VaeDecoder(Latent,vae,(Image)=>{

            SaveImage("path", image , ()=>{  
                //done
           })
        })

    })
  })
})
}
```

**Comment by omar92** [comment]
is there a possible way to convert current nodes to js functions that lloks like that and we excute them as graph

**Reply by KaruroChori** [reply]
https://github.com/KaruroChori/comfyui-bun-client
Sort of with external clients. But it is not well documented.

A workflow could be something like:
```ts
export const workflow = (opts: {
    positive: string, negative: string, multiplier?: number, denoise?: number, seed?: number, strength?: number
}, ctx?: Map<number, Node>) => {
    const comfy = Workflow(ctx);

    const node_2 = new comfy["CheckpointLoaderSimple"]({ "ckpt_name": "sd15/dreamshaper_8.safetensors" })
    const node_4 = new comfy["CLIPTextEncode"]({ "text": `${opts.positive} masterpiece, best quality, photograph`, "clip": node_2["CLIP"] })
    const node_5 = new comfy["CLIPTextEncode"]({ "text": `bad quality, bad picture, ${opts.negative}`, "clip": node_2["CLIP"] })
    //....
    const output_image = new comfy["PreviewImage"]({ "images": node_7["IMAGE"] })

    return { workflow: comfy, out: { colorized: output_image.$uid }, in: { src: "pasted/unique-name-0.png" } };
}

```

and gets used like:

```ts
const wf = workflow({ positive: "....", negative:  "...." })
const job = client.job(await wf.workflow.$compile(client.uid));
job.onStart(async () => {
    await $`spd-say -w "Processing ${trial[0]}"`
}).onError(async () => {
    await $`spd-say -w "Oh no... it failed"`
}).onCompleted(async () => {
    await $`spd-say -w "Done!"`
})

try {
    await job.upload_from_files([/*...*/])
    await job.schedule();
    await job.completion();
    await job.collect_to_files([/*...*/])
}
catch (e) {
    console.error((e as Error).message);
    await $`spd-say -w "Oh no... it failed!"`
}
```

**Reply by omar92** [reply]
thanks so much - it seems that the links is part of the python code -- i wished we could save results as variables and use them as we need but also makes sense as it will only be refrence to something stored in python

**Reply by omar92** [reply]
if i to be able to build something on top of that i need to be able to make it generate functions or classes (JS) for each node dynamicly

**Reply by omar92** [reply]
for the need to save stuff as variables maybe i can create some custom nodes as variables to hold these values (image variable , clip variable ,. .... etc)

**Reply by KaruroChori** [reply]
> thanks so much - it seems that the links is part of the python code -- i wished we could save results as variables and use them as we need but also makes sense as it will only be refrence to something stored in python

They are not in python, links are directly specified when nodes are defined, like ` "clip": node_2["CLIP"]` for `node_4`. The graph notation used by comfybun is basically a DAG where links are described inline for each node and topological order is followed.  
Not too easy to write, but the workflow code is usually generated automatically from a png with the UI workflow embedded and just adjusted for headless automation.

> if i to be able to build something on top of that i need to be able to make it generate functions or classes (JS) for each node dynamicly

You can have a look at my code, as that is what it does https://github.com/KaruroChori/comfyui-bun-client/blob/master/src/comfy-types-gen.ts . It connects to the client, ask for the json dictionary of all supported nodes, and generate typescript code to cover each type of node.

**Comment by ltdrdata** [comment]
Execution does not occur on the front end. The execution of the node takes place on the Python side.

Creating it with JavaScript instead of Python essentially means building a framework that generates prompt JSON. 
Is there a specific reason for trying to do that on the JavaScript side?

**Reply by omar92** [reply]
it will be like a script frontend that reuse all of comfy ui nodes + the flexibility of programming (you can make logic in easy way )

**Reply by omar92** [reply]
it can also be enhanced later so as example model name will show list of all avaliable checkpoints so you can select and it will autocomplete it

**Comment by blepping** [comment]
Easier to do it from the Python side and there are already several options:

1. https://github.com/Chaoses-Ib/ComfyScript
2. https://github.com/pydn/ComfyUI-to-Python-Extension

You don't necessarily have to use those, but they may be handy for creating some examples of what a Comfy graph would look like as a Python script.
